modules/packages/ArgumentParser.chpl

```diff
@@ -505,9 +609,186 @@ module ArgumentParser {
       _helpHandler.printHelp();
     }
 
-    // TODO: Implement automatic help message generation based on flags/opts etc
-    proc genHelp() {
-      _helpHandler._helpMessage = "internal generated message";
+    // sets the helpHandler's empty help message to generated message from
+    // the helpGenerator, or overwrites the help message if `message` is a string.
+    proc setHelp(message:?t=none) {
+      if isNothingType(t) && _helpHandler._helpMessage.isEmpty() {
+        var msg = _helpGenerator.generateHelp(binaryName=_binaryName);
+        _helpHandler._helpMessage = msg;
+      } else if isString(t) {
+        _helpHandler._helpMessage = message;
+      } else if !isNothingType(t) {
+          compilerError("Error setting help message: Expected string, received " + t:string);
+      }
+    }
+
+    // Notify the generator about the arguments defined in this parser
+    proc setArguments(argStack: map(string, ArgumentHandler)) {
+      _helpGenerator._setArguments(argStack);
+    }
+  }
+
+
+  pragma "no doc"
+  class HelpGenerator {
+
+    // create some sections to store the help text per section
+    var argsSection:section;
+    var optsSection:section;
+    var subSection:section;
+    // we need to know what argument handlers are defined in the parser
+    var _argStack: map(string, borrowed ArgumentHandler);
+
+    proc init() {
+      argsSection = new section();
+      optsSection = new section();
+      subSection = new section();
+    }
+
+    // represents two columns for argument, the name/flag and value, plus the
+    // help text
+    record element {
```

Comment:
I don't know if we have fixed them all but just a warning in the past we have seen a lot of bugs with nested classes / records

Reply:
eeek! I did not know that when I set these up. If they prove problematic I'll refactor them out, they aren't user facing so it shouldn't be an issue.